components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/util/ActionMgtAuditLoggerUtil.java

```diff
@@ -0,0 +1,324 @@
+     *
+     * @param operation Operation associated with the state change.
+     * @param actionId  ID of the action to be logged.
+     * @param action    Action object to be logged.
+     */
+    public void printActionAuditLog(Operation operation, String actionId, Action action) {
+
+        if (!LoggerUtils.isEnableV2AuditLogs()) {
+            return;
+        }
+        JSONObject data = createAuditLogEntry(actionId, action);
+        buildAuditLog(operation, data);
+
+    }
+
+    /**
+     * Print action audit log related to the operation by the action type and action ID.
+     *
+     * @param operation  Operation associated with the state change.
+     * @param actionType Type of the action to be logged.
+     * @param actionId   ID of the action to be logged.
+     */
+    public void printActionAuditLog(Operation operation, String actionType, String actionId) {
+
+        if (!LoggerUtils.isEnableV2AuditLogs()) {
+            return;
+        }
+        JSONObject data = createAuditLogEntry(actionType, actionId);
+        buildAuditLog(operation, data);
+    }
+
+    /**
+     * Build audit log using the provided data.
+     *
+     * @param operation Operation to be logged.
+     * @param data      data to be logged
+     */
+    private void buildAuditLog(Operation operation, JSONObject data) {
+
+        AuditLog.AuditLogBuilder auditLogBuilder = new AuditLog.AuditLogBuilder(getInitiatorId(),
+                LoggerUtils.getInitiatorType(getInitiatorId()),
+                LoggerUtils.Initiator.System.name(),
+                LoggerUtils.Target.Action.name(),
+                operation.getLogAction()).
+                data(jsonObjectToMap(data));
+        triggerAuditLogEvent(auditLogBuilder);
+    }
+
+    /**
+     * Create audit log data with action.
+     * This method expects all the action fields to be non-null/non-empty.
+     *
+     * @param action Action to be logged.
+     * @return audit log data.
+     */
+    private JSONObject createAuditLogEntry(Action action) {
+
+        JSONObject data = new JSONObject();
+        data.put(LogConstants.ACTION_TYPE_FIELD, action.getType());
+        data.put(LogConstants.ACTION_ID_FIELD, action.getId());
+        data.put(LogConstants.ACTION_NAME_FIELD, action.getName());
+        data.put(LogConstants.ACTION_DESCRIPTION_FIELD, action.getDescription());
+        data.put(LogConstants.ACTION_STATUS_FIELD, action.getStatus());
+        data.put(LogConstants.ENDPOINT_CONFIG_FIELD, getAllEndpointData(action.getEndpoint()));
+        return data;
+    }
+
+    /**
+     * Create audit log data with action and ID.
+     * This method expects null/empty action fields.
+     *
+     * @param actionId ID of the action to be logged.
+     * @param action   Action to be logged.
+     * @return audit log data.
+     */
+    private JSONObject createAuditLogEntry(String actionId, Action action) {
+
+        JSONObject data = new JSONObject();
```

Comment:
In this method, we are checking values are null and if not we add to the data. Why we need to have another createAuditLogEntry(Action action). IMO,  we can utilize the same method.

Reply:
`createAuditLogEntry(Action action) ` this method is expected to be used with ADD operation where each field should be non-null/non-empty. `createAuditLogEntry(String actionId, Action action)` this method is to be used with UPDATE operations where we expect null/empty fields by design.